include/crm/common/nvpair.h

```diff
@@ -23,16 +23,72 @@ extern "C" {
 #  include <libxml/tree.h>  // xmlNode
 #  include <crm/crm.h>
 
-typedef struct pcmk_nvpair_s {
-    char *name;
-    char *value;
-} pcmk_nvpair_t;
-
-GList *pcmk_prepend_nvpair(GList *nvpairs, const char *name, const char *value);
-void pcmk_free_nvpairs(GList *nvpairs);
-GList *pcmk_sort_nvpairs(GList *list);
-GList *pcmk_xml_attrs2nvpairs(xmlNode *xml);
-void pcmk_nvpairs2xml_attrs(GList *list, xmlNode *xml);
+/*
+ *  nvpairs (list of nvpair objects) encapsulating datatype
+ *  + basic public operations
+ */
+
+/* name that wouldn't need to be exposed if -fms-extensions CFLAG was a norm */
+#define pcmk_nvpairs_s _GSList
+struct pcmk_nvpairs_s;
+
+/*!
+ * \brief Opaque encapsulation of a list of name/value pairs
+ */
+typedef struct pcmk_nvpairs_s pcmk_nvpairs_t;
+
+/*!
+ * \brief Prepend a name/value pair to (a list of) nvpairs
+ *
+ * \param[in,out] nvpairs  List to modify
+ * \param[in]     name     New entry's name
+ * \param[in]     value    New entry's value
+ *
+ * \return New head (of list) of nvpairs
+ * \note The caller is responsible for freeing the nvpairs object
+ *       with \c pcmk_free_nvpairs().
+ */
+pcmk_nvpairs_t *pcmk_prepend_nvpair(pcmk_nvpairs_t *nvpairs, const char *name,
+                                    const char *value);
```

Comment:
Please don't convert in-file comments to header comments. The header files are not Pacemaker's API documentation; the online documentation at http://clusterlabs.org/pacemaker/doxygen/ is Pacemaker's API documentation, whether or not that is the favorite means of consumption of API documentation. API users are not the primary audience for Pacemaker source code; the only known API users outside Pacemaker itself are sbd and dlm_controld, and it is not expected that the API is useful to very many people. Instead, Pacemaker developers themselves are the primary audience of the code, and thus readability of the code is of much higher importance than helping API users. For that purpose, having the doxygen block with the code of the function is much, much more helpful when tracing bugs in the Pacemaker code than having to bounce back and forth between three code files (caller, called, and header).

Reply:
> Please don't convert in-file comments to header comments. The header
> files are not Pacemaker's API documentation;

This belongs to best practices, regardless of who consumes the API
(we are _not_ talking about private/hidden functions now).
Can we agree on that or are you going to stay in denial, requesting
that C files are always kept around when developing against the
pacemaker libraries?

> Readability of the code is of much higher importance than helping API
> users.

How readibility of the code relates to the documentation of functions,
types and structs?
The most readable code is that which doesn't require any documentation.

> having to bounce back and forth between three code files (caller,
> called, and header).

There are ways to conveniently display excerpts of various files on
one screen at the same time.  There are also various ways to navigate
between declarations and definitions conveniently in various editors
and IDEs.  Are we going to optimize for those giving up on these
common coding practices at the cost non-standard approach?

Reply:
> > Please don't convert in-file comments to header comments. The header
> > files are not Pacemaker's API documentation;
> 
> This belongs to best practices, regardless of who consumes the API
> (we are _not_ talking about private/hidden functions now).
> Can we agree on that or are you going to stay in denial, requesting
> that C files are always kept around when developing against the
> pacemaker libraries?

It's not denial, it's a statement of policy: header files are not pacemaker's API documentation. We have API documentation, it's on clusterlabs.org.

But there is NO ONE "developing against the pacemaker libraries". A public API was created in the early days before anyone knew how things would actually be used, and we're stuck with a certain degree of backward compatibility guarantees (which thankfully can be broken at select times). In practice, the only known software to use the pacemaker API (outside pacemaker itself) is sbd and dlm_controld. The sbd developers are all also pacemaker developers, and the dlm_controld developers want as little to do with pacemaker as possible. I guarantee you none of them are looking for pacemaker API documentation.

What outside users do want from a pacemaker API (high-level functions for managing cluster services) is currently provided by pacemaker's command line tools. The C API is neither intended nor suited for this purpose.

As such, the primary intended audience of the doxygen comment blocks is not the mythical external API users, but the very real pacemaker developers who have to look at the code every day.

> > Readability of the code is of much higher importance than helping API
> > users.
> 
> How readability of the code relates to the documentation of functions,
> types and structs?

It is far, far more readable to have the documentation of a thing with its definition.

A typical workflow for me involves tracing through pacemaker detail logs for a suspect incident to determine whether/where it is a bug. This generally involves many levels of abstraction and function depth, bouncing back and forth between numerous files, drilling down through levels of function calls, and almost always requiring a set of notes to follow the logic. Being able to read what a function's arguments are with the function definition is an enormous time-saver to me. I don't need to be slowed down for the benefit of a nonexistent audience.

> The most readable code is that which doesn't require any documentation.

A noble goal but rarely obtainable in the best of cases, much less the pacemaker code base. It is not realistic to suggest that a function and its arguments can be named in a way that provides all necessary information about how the function should be used. Some commenting is required.

> > having to bounce back and forth between three code files (caller,
> > called, and header).
> 
> There are ways to conveniently display excerpts of various files on
> one screen at the same time. There are also various ways to navigate
> between declarations and definitions conveniently in various editors
> and IDEs. Are we going to optimize for those giving up on these
> common coding practices at the cost non-standard approach?

And for the pacemaker project, we've chosen doxygen for that purpose, and its result is kept on the clusterlabs.org website. That's the standard approach we've taken, and anyone can take advantage of it. Doxygen doesn't care whether the comment block is with the declaration or the definition, but I do.